### ISSUE TITLE
Contacts are shown offline when they are online

### ISSUE THREAD

**Issue by sagb** [irungentoo/toxcore#1349]
Some friends aren't shown as online while they are for hours (we can even communicate via groupchat). People in official groupchat suggest that this is toxcore issue.
I use qTox from latest dpkg package (1.0-1703), probably core version from Apr 26 2015, "offline" people use qTox and Toxic.
This MAY be fixed by restarting client, but not always.

**Comment by LittleVulpix** [comment]
I noticed this happened to me when I was in a tcp-only mode and my friends were in udp mode, although I thought this shouldn't affect anything; but I was unable to see my friend online, and when I switched back to udp mode, it worked just fine straight away. It generally happens with friends who are far away or have poor connection.

**Comment by zetok** [comment]
@LittleVulpix just to clarify; have those friends not come online for hours in your client (even for a brief moment), even though they were online and were able to communicate with other people?

**Comment by LittleVulpix** [comment]
@zetok In one case, the friend only has me in their tox, so it's hard to say, but in the other case, yes, that was exactly how it was. I even have a second tox account, which is how I noticed. I logged into the second tox account on a different computer in my home network and was surprised to see 2 extra people online. ( the second account had utox with udp enabled) So I switched udp on in my main account and reconnected and I could see everyone online properly, just like on that secondary account.

**Comment by aaannndddyyy** [comment]
@sagb: yes, this is a known bug that has  existed for a long time now. But in my experience it has gotten a lot better.
Mainly with people on poor connections.

**Comment by sagb** [comment]
@LittleVulpix: me and my friends never changed default UDP mode to TCP. So I believe the full reconnect (with new bootstrap nodes) fixed the issue for you, as well as for me, not switching between TCP and UDP.

**Comment by sagb** [comment]
aaannndddyyy, are there plans for fixing, or it's left as is?

**Comment by LittleVulpix** [comment]
@sagb yep, I think so too, but it's difficult to verify because this only happens "sometimes".
@aaannndddyyy :D I'd appreciate a fix as well, hehehe. But I guess it's more of an @irungentoo thing to fix? Either way, it is an annoying problem if you sometimes can't see people online when they are...

**Comment by aaannndddyyy** [comment]
Yep, not my call. But of course, bug are supposed to be fixed - what a question. However, for fixing it, the root cause must be known...
And once you are connected to your friends, they don't randomly disappear, at least not here, so it is related to the onion, it seems

**Comment by aaannndddyyy** [comment]
re TCP and UDP, I can confirm that it also happens if both parties are on TCP-only.

**Comment by LittleVulpix** [comment]
I think this requires a lot more attention, seeing as it's very problematic to not see someone online and think they're not, when they are online on their side and they're thinking the exact same thing.

It's one of the most basic things to be able to do on a chat program - see each other. Without that, you can't chat (on tox); faux offline messaging is not a solution.

When someone is online, I want to be able to see them - OR they must know that they are not online.

**Comment by irungentoo** [comment]
If you could say which clients you and your contacts are using it would be useful to debug the issue.

**Comment by aaannndddyyy** [comment]
Antox-Antox, Antox-utox, utox-utox.

**Comment by LittleVulpix** [comment]
For me, it's miranda NG ( http://www.miranda-ng.org/en/ ) to uTox or qtox for a friend who has Mac instead of a PC.

**Comment by irungentoo** [comment]
If you encounter the bug with a client that isn't utox, look in the code of the client if they use: tox_add_tcp_relay(). If they don't, then tell the dev to use that function, if they do, report the bug here.

**Comment by aaannndddyyy** [comment]
what about utox? does it use that function too, or why does utox have this issue too?

**Comment by irungentoo** [comment]
uTox uses that function correctly.

**Comment by aaannndddyyy** [comment]
then it seems that that function is not the problem and the problem lies somewhere else.

**Comment by LittleVulpix** [comment]
@irungentoo thanks! I asked our dev to take a look at this. Although, as @aaannndddyyy says, it happens with uTox to uTox as well (though I am unable to check this since I don't actually use uTox as my primary chat program for tox.

**Comment by sagb** [comment]
@irungentoo, I have this situation routinely with qTox <-> qTox.

This prevents my intention to present tox as "better" skype replacement, because its worse reliability is too obvious for users.

tox_add_tcp_relay is mentioned in qTox in src/core/core.cpp and probably called as: Core::start -> Core::process -> Core::bootstrapDht -> tox_add_tcp_relay.

**Comment by irungentoo** [comment]
@sagb that function was added to qtox less than a day ago.

**Comment by LittleVulpix** [comment]
@irungentoo our dev looked at tox_add_tcp_relay and added some logging to see if there is a problem with it, but I couldn't see a problem. I have a question; this is a function you call at bootstrap? Because if that's the case, it wouldn't really be affecting what happens later on, no? I mean, the situation is that I log in, I see everyone; but over time, somewhere, along the way, I lose connectivity to several of my contacts. If then, at any point, I reconnect (go offline and back online), I see everyone again.

I will try to wait for you on IRC to see if you could perhaps spare a bit of your time and help me highlight what my dev needs to focus on, because it's happening somewhat consistently now :/

**Comment by udevd** [comment]
I have the same issue as OP, expect for the qtox version: I have 1.0-1737 (it's up-to-date at the moment).

**Comment by udevd** [comment]
...or perhaps should I say `debian–up–to–date'.

**Comment by thiras** [comment]
Confirmed with latest qtox on windows (cfeeb03c5ba1da034ebd6b3d04c05972c7a8ad8f - pr1968) with latest official debian/ubuntu repos qtox. Solid connection, UDP enabled on windows.

Linux qtox (at default settings) had very bad connection (especially at upload) and i was the only and first contact on its contact list.

Restarting both of clients solved problem.

**Comment by ProMcTagonist** [comment]
It's happened often for me with qTox on Debian <> qTox on Windows for a long time. Reconnecting on either side fixes it.

**Comment by zetok** [comment]
I had experienced this problem today, using up-to-date qTox with a toxcore that is 6 days old. Other contact used µTox 0.3.2, and after they restarted their client I was able to talk with them.

**Comment by aaannndddyyy** [comment]
it is still happening with latest core and both contacts on µtox

**Comment by LittleVulpix** [comment]
Yep, I can confirm. Also happens on miranda and qtox.

**Comment by Impyy** [comment]
This happened to me today as well, using latest toxcore and not using tcp-only mode.

**Comment by LittleVulpix** [comment]
This happened to me again today, using TCP-only mode. Receiving a file from a contact with poor connection caused him to show offline to me temporarily, interrupting and failing the transfer as a consequence.

**Comment by aaannndddyyy** [comment]
Still happens. Only fix is to reconnect.

**Comment by GrayHatter** [comment]
I'm seeing this a lot if there's any kind of issue with and A/V call from uTox on devel to uTox on devel, both on the same LAN, usually one or both clients need a restart. And according to clever in #tox-dev it's happening to him LAN client to LAN client in toxvpn as well.

**Comment by GrayHatter** [comment]
@irungentoo

**Comment by cleverca22** [comment]
I've noticed the same thing happening when both ends are capable of UDP

**Comment by aaannndddyyy** [comment]
Happens to me regularly without lan or a/v being involved. My contacts are mostly in another country.

**Comment by fcore117** [comment]
Happens still with both TCP and UDP modes for me.

**Comment by GrayHatter** [comment]
can someone duplicate this and acquire a toxcore log?

**Comment by zetok** [comment]
>can someone duplicate (…)

![still reproducible](https://cloud.githubusercontent.com/assets/3148759/12678406/6ce77ede-c696-11e5-9f8d-fc3ba5f968e2.png)
↑ still reproducible

**Comment by GrayHatter** [comment]
@zetok the point was to get a debug log from toxcore. So that we can find out why toxcore can't "see" the peer.

**Comment by zetok** [comment]
@GrayHatter then I would suggest you to get one and analyse it to find out why it's not working, since it's reproducible.

Unless you're asking to upload my log ↓ somewhere? Are there even services that would host such file? Are you sure that you want to go through years of logs?
```
$ du -h libtoxcore.log 
24G	libtoxcore.log
```

**Comment by qasmoke** [comment]
https://github.com/tux3/qTox/issues/1560#issuecomment-220223806

How about:
1. Ping to DHT sever every 1 minutes, if the user USE PROXY and TCP ONLY(not udp).
2. If the ping failed, try again immediately.
3. If failed again, initiate restart sequence.

This bug is very annoying, please fix it...

**Comment by qasmoke** [comment]
^ PING = not about ICMP, but send random small packet to DHT server via TCP over proxy.

**Comment by LittleVulpix** [comment]
Well I mean when you have tox running, there is a constant stream of data as it partakes in DHT communication. And by constant I mean sometimes up to 10-20kB/s.

**Comment by ddobrev** [comment]
@irungentoo this bug is more than a year old now and is the cause of the most frustrating problem in clients (such as qTox). When is it going to be fixed?

**Comment by aaannndddyyy** [comment]
tox seems to be dead

**Comment by GrayHatter** [comment]
Lol @optimumtact another for the list

**Comment by ddobrev** [comment]
@GrayHatter you could not just call him but prove him wrong by fixing the bug. :)

**Comment by ddobrev** [comment]
Has this one perhaps been fixed in the new branch?

**Comment by proxifier** [comment]
This issue can be fixable already.
Sinply add these settings into options, and let users use it.

[X] Reconnect Automatically Every (pull down menu V )
1 hours
2 hours
3 hours
6 hours
12 hours
1 day

> if ($qTox_running_time >= 86400) then
>  qtox_reconnect(); // Same action when user press proxy 'reconnect' button in qtox settings
> fi

**Comment by proxifier** [comment]
And

[X] Do not reconnect if the user has open chat window
(e.g., if the user using multiple window mode, do not reconnect if there is any chat window opened)

**Comment by vladimiry** [comment]
@proxifier 

> Reconnect Automatically

Could it also do auto-reconnect when computer goes from the hibernate/sleep? When I placed the original issue, a few years ago (yes long time ago), I noticed that issue happens after hibernate/sleep mode.

**Comment by aaannndddyyy** [comment]
@proxifier:
That's an ugly-hack work-around, not a fix.

**Comment by ddobrev** [comment]
if it's a hack, what's the proper solution?

**Comment by aaannndddyyy** [comment]
The proper solution is reconnecting at least every two minutes… ;)

**Comment by proxifier** [comment]
@aaannndddyyy 

I hope you're not offending me, right bud?

@vladimiry

yes, we also need to consider "hibernate" "sleep" situation.
So, the code will be look like...

> $qtox_config_TTL=86400; // 1day as seconds
> void detect_hibernateNsleep(){......}
> if ($qTox_running_time >= $qtox_config_TTL){
> if (qTox_is_not_uploadNdownload_files() && qTox_has_not_open_chat_window()){
> qtox_reconnect();
> }
> }

**Comment by aaannndddyyy** [comment]
@proxifier:
No.
And I hope you're not offending me either, because that would not be 
okay.

**Comment by GrayHatter** [comment]
Well, you're on the toxcore repo... So it'll have nothing to do with qTox...

But the proper solution would be figure out why toxcore doesn't try to
reconnect to these offline friends.

**Comment by aaannndddyyy** [comment]
If it helps diagnosing the problem, I can tell that it does not affect 
all contacts at once, but only one or two at a time. Sometimes none.
And yes, I have, of course, verified that those one or two offline 
contacts are really online.

**Comment by ddobrev** [comment]
I am really happy this is finally getting some attention. Keep up the good work, fellows.

**Comment by GrayHatter** [comment]
Do you appear offline to them?

**Comment by aaannndddyyy** [comment]
Yes, it's mutual.

**Comment by ddobrev** [comment]
I can see the development of toxcore has resumed, there was a commit in the "new" branch yesterday. Does that mean this problem is going to be checked soon?

**Comment by vladimiry** [comment]
Just tested a case using latest q-tox clients on Win 10 and Archlinux. After getting from the sleep mode (on the next day) both users see each other as an offline contacts. A few years ago it worked in the same way.

**Comment by ddobrev** [comment]
Does anybody work on this bug? Does anyone work on toxcore at all?

**Comment by Spagy** [comment]
I don't know about work on this bug specifically but work on toxcore has moved to [TokTok/toxcore](https://github.com/TokTok/toxcore). See [the blog](https://blog.tox.chat/2016/08/update-new-client-xenial-packages-tox-in-google-play-toxcore-fork-and-more/) for an explanation.

**Comment by iphydf** [comment]
Do you have specific steps to reproduce this issue or does it randomly occur? Is it related to the computer going to sleep and waking up? If so, can it be reproduced by cutting the network connection for a while?

**Comment by ddobrev** [comment]
Hello @iphydf . Sorry, I can remember now you did tell me this repository had been abandoned and your fork was the real one instead. To reproduce the problem, it's enough for the Internet connection to disappear for a short while - when it's turned off and back on, when a VPN is activated or deactivated, when the computer is put to sleep and then woken up. The bug may not occur every time but it will happen more often than not.

**Comment by MikeHansenHH** [comment]
Bug still exists with uTox 0.9.8 for OS X. :-(

**Comment by ovalseven8** [comment]
@MikeHansenHH: Please repost [here](https://github.com/TokTok/c-toxcore/issues/237) because this repository is more or less dead.
